wordpress.org/public_html/wp-content/plugins/wporg-gp-customizations/inc/cli/class-detect-duplicate-translations.php

```diff
@@ -0,0 +1,286 @@
+			$results = $wpdb->get_results( $prepared_query, ARRAY_A );
+
+			if ( $wpdb->last_error ) {
+				// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+				echo 'Error in the SQL query: ' . $wpdb->last_error;
+				$this->last_sql_error = $wpdb->last_error;
+				break;
+			}
+
+			$duplicate_entries = array_merge( $duplicate_entries, $results );
+
+			if ( $this->print_sql ) {
+				foreach ( $results as $result ) {
+					// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+					echo "- Translation_set_id: {$result['translation_set_id']}, Original_id: {$result['original_id']} -> ";
+					// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+					echo "SELECT * FROM `{$wpdb->gp_translations}` WHERE `original_id` = {$result['original_id']} AND `translation_set_id` = {$result['translation_set_id']}\n";
+				}
+			}
+		}
+		if ( $this->verbose ) {
+			// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+			echo 'Number of duplicate entries: ' . number_format_i18n( count( $duplicate_entries ) ) . "\n";
+		}
+		return $duplicate_entries;
+	}
+
+	/**
+	 * Update the duplicated entries.
+	 *
+	 * Set all translations to old except the last one, updating the date_modified and user_id_last_modified fields.
+	 *
+	 * @return void
+	 */
+	private function update_duplicates() {
+		global $wpdb;
+
+		foreach ( $this->duplicates as $duplicate ) {
+			$update_query = $wpdb->prepare(
+				"UPDATE {$wpdb->gp_translations}
+				SET status = CASE
+					WHEN id = (
+						SELECT id
+						FROM (
+							SELECT id
+							FROM {$wpdb->gp_translations}
+							WHERE original_id = %d
+							AND translation_set_id = %d
+							AND status = 'current'
+							ORDER BY date_modified DESC, id DESC
+							LIMIT 1
+						) AS subquery
+					) THEN 'current'
+					ELSE 'old'
+				END,
+				user_id_last_modified = CASE
+					WHEN id = (
+						SELECT id
+						FROM (
+							SELECT id
+							FROM {$wpdb->gp_translations}
+							WHERE original_id = %d
+							AND translation_set_id = %d
+							AND status = 'current'
+							ORDER BY date_modified DESC, id DESC
+							LIMIT 1
+						) AS subquery
+					) THEN user_id_last_modified
+					ELSE NULL
+				END,
+				date_modified = CASE
+					WHEN id = (
+						SELECT id
+						FROM (
+							SELECT id
+							FROM {$wpdb->gp_translations}
+							WHERE original_id = %d
+							AND translation_set_id = %d
+							AND status = 'current'
+							ORDER BY date_modified DESC, id DESC
+							LIMIT 1
+						) AS subquery
+					) THEN date_modified
+					ELSE NOW()
+				END
+				WHERE original_id = %d
+				AND translation_set_id = %d
+				AND status = 'current';
+			",
+				$duplicate['original_id'],
+				$duplicate['translation_set_id'],
+				$duplicate['original_id'],
+				$duplicate['translation_set_id'],
+				$duplicate['original_id'],
+				$duplicate['translation_set_id'],
+				$duplicate['original_id'],
+				$duplicate['translation_set_id']
+			);
```

Comment:
Maybe I am misinterpreting what the code does but wouldn't this be much simpler?
```suggestion
foreach ( $this->duplicates as $duplicate ) {
	$id = $wpdb->get_var(
		$wpdb->prepare(
			"SELECT id
			FROM {$wpdb->gp_translations}
			WHERE original_id = %d
			AND translation_set_id = %d
			AND status = 'current'
			ORDER BY date_modified DESC, id DESC
			LIMIT 1",
			$duplicate['original_id'],
			$duplicate['translation_set_id']
		)
	);

	$update_query = $wpdb->prepare(
		"UPDATE {$wpdb->gp_translations}
			SET status = CASE
				WHEN id = %d THEN 'current'
				ELSE 'old'
			END,
			user_id_last_modified = CASE
				WHEN id = %d THEN user_id_last_modified
				ELSE NULL
			END,
			date_modified = CASE
				WHEN id = %d THEN date_modified
				ELSE NOW()
			END
			WHERE original_id = %d
			AND translation_set_id = %d
			AND status = 'current';
		",
		$id,
		$id,
		$id
		$duplicate['original_id'],
		$duplicate['translation_set_id']
	);
```

Reply:
Or even just:
```suggestion
foreach ( $this->duplicates as $duplicate ) {
	$id = $wpdb->get_var(
		$wpdb->prepare(
			"SELECT id
			FROM {$wpdb->gp_translations}
			WHERE original_id = %d
			AND translation_set_id = %d
			AND status = 'current'
			ORDER BY date_modified DESC, id DESC
			LIMIT 1",
			$duplicate['original_id'],
			$duplicate['translation_set_id']
		)
	);

	$update_query = $wpdb->prepare(
		"UPDATE {$wpdb->gp_translations}
			SET status = 'old',
			user_id_last_modified = NULL,
			date_modified = NOW()
			WHERE original_id = %d
			AND translation_set_id = %d
			AND id != %d
			AND status = 'current';
		",
		$duplicate['original_id'],
		$duplicate['translation_set_id'],
		$id
	);
```

Reply:
I have updated the code with this improvement!